auth/auth_test.go

```diff
@@ -1240,3 +1240,933 @@ func TestAuthenticateUntrustedJWT(t *testing.T) {
+		roleChannels: map[string]ch.TimedSet{},
+	}
+
+	auth := NewAuthenticator(testBucket, &testMockComputer)
+	aliceUserPrincipal, fooPrincipal := initializeScenario(t, auth)
+
+	testMockComputer.addRoleChannels(t, auth, fooPrincipal, "foo", "ch1", 5)
+	testMockComputer.addRole(t, auth, aliceUserPrincipal, "alice", "foo", 20)
+
+	// Get Principals / Rebuild Seq 25
+	aliceUserPrincipal, fooPrincipal = getPrincipals(t, auth)
+
+	// Ensure user can see ch1 (via role)
+	// Verify history
+	assert.ElementsMatch(t, []string{"!", "ch1"}, fooPrincipal.Channels().AllChannels())
+	assert.True(t, aliceUserPrincipal.CanSeeChannel("ch1"))
+	assert.Equal(t, 0, len(aliceUserPrincipal.RoleHistory()))
+	assert.Equal(t, 0, len(aliceUserPrincipal.ChannelHistory()))
+	assert.Equal(t, 0, len(fooPrincipal.ChannelHistory()))
+
+	testMockComputer.removeRole(t, auth, aliceUserPrincipal, "alice", "foo", 45)
+
+	// Get Principals / Rebuild Seq 50
+	aliceUserPrincipal, fooPrincipal = getPrincipals(t, auth)
+
+	// Ensure user cannot see ch1 (via role)
+	// Verify history
+	assert.False(t, aliceUserPrincipal.CanSeeChannel("ch1"))
+	userRoleHistory, ok := aliceUserPrincipal.RoleHistory()["foo"]
+	require.True(t, ok)
+	assert.Equal(t, GrantHistorySequencePair{Seq: 20, EndSeq: 45}, userRoleHistory.Entries[0])
+	assert.Equal(t, 0, len(aliceUserPrincipal.ChannelHistory()))
+	assert.Equal(t, 0, len(fooPrincipal.ChannelHistory()))
+
+	testMockComputer.removeRoleChannel(t, auth, fooPrincipal, "foo", "ch1", 55)
+	testMockComputer.addRole(t, auth, aliceUserPrincipal, "alice", "foo", 65)
+	testMockComputer.addRoleChannels(t, auth, fooPrincipal, "foo", "ch1", 75)
+
+	// Get Principals / Rebuild Seq 80
+	aliceUserPrincipal, fooPrincipal = getPrincipals(t, auth)
+
+	// Ensure user can see ch1 (via role)
+	// Verify history
+	assert.True(t, aliceUserPrincipal.CanSeeChannel("ch1"))
+	assert.Equal(t, GrantHistorySequencePair{Seq: 20, EndSeq: 45}, userRoleHistory.Entries[0])
+	assert.Equal(t, 0, len(aliceUserPrincipal.ChannelHistory()))
+	assert.Equal(t, 0, len(fooPrincipal.ChannelHistory()))
+
+	testMockComputer.removeRoleChannel(t, auth, fooPrincipal, "foo", "ch1", 85)
+	testMockComputer.removeRole(t, auth, aliceUserPrincipal, "alice", "foo", 95)
+
+	// Get Principals / Rebuild Seq 110
+	aliceUserPrincipal, fooPrincipal = getPrincipals(t, auth)
+
+	// Ensure user cannot see ch1 (via role)
+	// Verify history
+	assert.False(t, aliceUserPrincipal.CanSeeChannel("ch1"))
+	userRoleHistory, ok = aliceUserPrincipal.RoleHistory()["foo"]
+	require.True(t, ok)
+	assert.Equal(t, GrantHistorySequencePair{Seq: 20, EndSeq: 45}, userRoleHistory.Entries[0])
+	assert.Equal(t, GrantHistorySequencePair{Seq: 65, EndSeq: 95}, userRoleHistory.Entries[1])
+
+	channelHistory, ok := fooPrincipal.ChannelHistory()["ch1"]
+	require.True(t, ok)
+	assert.Equal(t, GrantHistorySequencePair{Seq: 75, EndSeq: 85}, channelHistory.Entries[0])
+
+	assert.Equal(t, 0, len(aliceUserPrincipal.ChannelHistory()))
+}
+
+// Scenario 3
```

Comment:
I'm not sure these descriptions are that illuminating as written.  I don't know what 'Get Doc' means in the context of this test, and I think the main difference between scenarios is timing of requests (compared to the grants/revocation), and that's not described.

Reply:
If you want to list the steps (as opposed to a verbal description of what differentiates the scenario), it might be better in the form:
```
[describe setup]
- changes request, expected result
[describe what's changed]
- changes request, expected result
[describe what's changed]
- changes request, expected result
```
repeat as needed

Reply:
Also, the sequence for role/channel grants/revocations is the same for all these scenarios, is that right?  The only thing changing is the point at which you're simulating a replication?  If so, that sounds like something that would actually make more sense as a single unit test, with an 'ensure user can...' block for each scenario.

Reply:
Yes. Its true that the same grants and revocations occur in each scenario. However, due to how the history is built you can't really do this in one test. If you mean do it as a case table kind of test that would be possible, I considered it, but it got difficult to read whats going on exactly in each step.

Reply:
I thought the role/channel history should be identical for all the scenarios?  My reading of the spreadsheet is that the user/role/channel revocation history is the same across all scenarios, and the differences are when a client attempts to replicate.  Can you explain more what you mean when you say 'due to how the history is built, you can't really do this in one test'?  If it's due to information being persisted to the user document on authentication, you could invalidate channels/roles prior to doing that check.  However, you'd also want to test to ensure that invalidation isn't required (i.e. one version of the test makes sure the history is as expected without validation, one checks with validation.

Reply:
Added comments like above and other issues should be addressed as per discussion